components/application-mgt/org.wso2.carbon.device.application.mgt.core/src/main/java/org/wso2/carbon/device/application/mgt/core/impl/SubscriptionManagerImpl.java

```diff
@@ -142,32 +100,52 @@
         return null;
     }
 
-    private List<DeviceIdentifier> installApplication(String applicationUUID, List<DeviceIdentifier> deviceList)
+    private List<DeviceIdentifier> installApplication(String applicationUUID, List<DeviceIdentifier> deviceList, String versionName)
             throws ApplicationManagementException {
         List<DeviceIdentifier> failedDeviceList = new ArrayList<>(deviceList);
-        List<org.wso2.carbon.device.mgt.common.DeviceIdentifier> activeDeviceList = new ArrayList<>();
+        // Todo: try whether we can optimise this by sending bulk inserts to db
+        // Todo: atomicity is not maintained as deveice managment provider service uses separate db connection. fix this??
+        log.info("Install application: " + applicationUUID + "[" + versionName + "]" + " to: " + deviceList.size() + " devices.");
         for (DeviceIdentifier device : deviceList) {
             org.wso2.carbon.device.mgt.common.DeviceIdentifier deviceIdentifier = new org.wso2.carbon.device.mgt
                     .common.DeviceIdentifier(device.getId(), device.getType());
             try {
-                DeviceManagementDAOFactory.openConnection();
-                if (DeviceManagementDAOFactory.getDeviceDAO().getDevice(deviceIdentifier).isEmpty()) {
-                    log.error("Device with ID: " + device.getId() + " not found to install the application.");
+                DeviceManagementProviderService dmpService = HelperUtil.getDeviceManagementProviderService();
+                if (!dmpService.isEnrolled(deviceIdentifier)) {
+                    log.error("Device with ID: [" + device.getId() + "] is not enrolled to install the application.");
                 } else {
                     if (log.isDebugEnabled()) {
-                        log.debug("Prepare application install to : " + device.getId());
+                        log.debug("Installing application to : " + device.getId());
                     }
-                    activeDeviceList.add(deviceIdentifier);
-                    DAOFactory.getSubscriptionDAO().addDeviceApplicationMapping(device.getId(), applicationUUID, false);
+                    //Todo: generating one time download link for the application and put install operation to device.
+
+                    // put app install operation to the device
+                    ProfileOperation operation = new ProfileOperation();
+                    operation.setCode("INSTALL_APPLICATION");
```

Comment:
Let's add a constant for the install application operation code.

Reply:
+1

Reply:
Updated with: https://github.com/wso2/carbon-device-mgt/pull/1032/commits/659713b59348bff41ba1fda1a990b02f425fcd2b